### ISSUE TITLE
BinActive::forward() does't seem to match paper

### ISSUE THREAD

**Issue by honglh** [jiecaoyu/XNOR-Net-PyTorch#100]
Code under discussion
 ```
  # https://github.com/jiecaoyu/XNOR-Net-PyTorch/blob/master/CIFAR_10/models/nin.py#L9
  # BinActive.forward()
```
```
   # https://github.com/jiecaoyu/XNOR-Net-PyTorch/blob/master/CIFAR_10/models/nin.py#L43
   #BinConv2d.forward()
```
`(*)` means `conv` operation, `o` is element-wise product 

The paper's algorithm for Input binarization is:  
```
I * W ~= (sign(I) (*) sign(W)) o Ka = ((sign(I) (*) sign(W)) o (A (*) k)a
where
A =  torch.mean(input.abs(), 1, keepdim=True)
k  = an averaging kernel with value 1/(w*h)
```
so
`A(*)k` is to averaging each input element with its neighboring elements. This is ***missing*** in the current implementation, where only `(sign(I) (*) sign(W)) o a` is calculated.
```
x, mean = BinActive()(x) 
x = self.conv(x)
```

To capture the convolution with `A` and `k` from the paper, I  would expect pseudo code like this in `BinConv2d.forward()`
```
x, mean = BinActive()(x) 
# <=== === === === === === === === === === START
A = mean #shape N, 1, W, H
sign_I = x #shape  N, Cin, W, H
kH = self.conv.weight.shape[2]
kW = self.conv.weight.shape[3]
k = torch.ones(1, 1, kH, kW) * (1/(kH*kW))
conv_Ak = torch.nn.Conv2d(1, 1, kH, kW, padding=(kH//2, kW//2))
conv_Ak.weight.data = k
K = conv_Ak(A) #shape N, 1, W, H

#now calculate sign_I (*)sign_W o Ka
# since self.conv.weight is already binarized by binarizeConvParams() before batch starts, 
# the `a` in `Ka` is included in `self.conv(x)` . The only missing part is `mul(K)`
# Hence:
x = self.conv(x).mul(K) 
# <=== === === === === === === === === === END
```

Can you check if my understanding of the discrepancy is correct?

**Comment by jiecaoyu** [comment]
Hi @honglh , yes, my implementation is different from the paper. This because computing the average values for inputs will make the training extremely slow. Also, it will not affect the accuracy we can get. Therefore, I remove that part from my implementation. I think the original implementation from the author also did not include that part.
